persistence/internal/instrumented/keyvalue.go

```diff
@@ -0,0 +1,303 @@
+package instrumented
+
+import (
+	"context"
+
+	"github.com/dogmatiq/veracity/internal/telemetry"
+	"github.com/dogmatiq/veracity/persistence/kv"
+	"go.opentelemetry.io/otel/metric"
+)
+
+// KeyValueStore is a decorator that adds instrumentation to a [kv.Store].
+type KeyValueStore struct {
+	Next      kv.Store
+	Telemetry *telemetry.Provider
+}
+
+// Open returns the keyspace with the given name.
+func (s *KeyValueStore) Open(ctx context.Context, name string) (kv.Keyspace, error) {
+	r := s.Telemetry.New(
+		"github.com/dogmatiq/veracity/persistence",
+		"keyspace",
+		telemetry.Type("store", s.Next),
+		telemetry.String("handle", handleID()),
+		telemetry.String("name", name),
+	)
+
+	ctx, span := r.StartSpan(ctx, "keyspace.open")
+	defer span.End()
+
+	next, err := s.Next.Open(ctx, name)
+	if err != nil {
+		span.Error("could not open keyspace", err)
+		return nil, err
+	}
+
+	ks := &keyspace{
+		Next:      next,
+		Telemetry: r,
+		OpenCount: r.Int64UpDownCounter(
+			"open_keyspaces",
+			metric.WithDescription("The number of keyspaces that are currently open."),
+			metric.WithUnit("{keyspace}"),
+		),
+		DataIO: r.Int64Counter(
+			"io",
+			metric.WithDescription("The cumulative size of the keys and values that have been read and written."),
+			metric.WithUnit("By"),
+		),
+		PairIO: r.Int64Counter(
+			"pair.io",
+			metric.WithDescription("The number of key/value pairs that have been read and written."),
+			metric.WithUnit("{pair}"),
+		),
+		KeySize: r.Int64Histogram(
+			"key.size",
+			metric.WithDescription("The sizes of the keys that have been read and written."),
+			metric.WithUnit("By"),
+		),
+		ValueSize: r.Int64Histogram(
+			"value.size",
+			metric.WithDescription("The sizes of the value that have been read and written."),
```

Comment:
Should `sizes` and `value` be plural matched here?
> sizes of the value

Reply:
Nice catch, as always :D  Fixed.